snapcraft/internal/remote_build/_launchpad.py

```diff
@@ -0,0 +1,227 @@
+# -*- Mode:Python; indent-tabs-mode:nil; tab-width:4 -*-
+#
+# Copyright (C) 2019 Canonical Ltd
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License version 3 as
+# published by the Free Software Foundation.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+import logging
+import os
+import shutil
+import time
+import urllib.request
+import urllib.error
+
+from lazr import restfulclient
+from launchpadlib.launchpad import Launchpad
+from typing import Any, Dict, List
+from xdg import BaseDirectory
+from snapcraft import InfoFile
+from snapcraft.project import Project
+from .errors import (
+    NoLaunchpadUsernameError,
+    RemoteBuilderNotReadyError,
+    RemoteBuildNotFoundError,
+)
+
+import snapcraft
+
+_LP_POLL_INTERVAL = 30
+_LP_SUCCESS_STATUS = "Successfully built"
+_LP_FAIL_STATUS = "Failed to build"
+
+logger = logging.getLogger(__name__)
+
+
+class LaunchpadClient:
+    """Launchpad remote builder operations."""
+
+    def __init__(self, project: Project, build_id: str) -> None:
+        self._id = build_id
+        self._core_channel = "stable"
+        self._snapcraft_channel = "edge"
+        self._name = project.info.name
+        self._waiting = []  # type: List[str]
+        self._data_dir = os.path.join(
+            BaseDirectory.save_data_path("snapcraft"), "launchpad"
+        )
+        self._cache_dir = os.path.join(
+            BaseDirectory.save_cache_path("snapcraft"), "launchpad"
+        )
+
+        os.makedirs(self._data_dir, mode=0o700, exist_ok=True)
+        self._credentials = os.path.join(self._data_dir, "credentials")
+
+    def login(self, user: str) -> None:
+        info = InfoFile(os.path.join(self._data_dir, "config.yaml"))
+        if user:
```

Comment:
I would prefer this to look more like the config class we have for error handling and other preferences.

We already have credentials config available, maybe we just need to add the launchpad stuff there and not require Info

Reply:
Changed configuration data storage to use the existing infrastructure used for credentials, under a new "Launchpad" section.